src/python/pants/engine/README.md

```diff
@@ -17,63 +17,69 @@ Once the engine is instantiated with a valid set of `@rule`s, a caller can synch
 
 #### Example
 
 As a very simple example, you might register the following `@rule` that can compute a `String`
-Product given a single `Int` input.
+Product given a single `Int` argument.
 
 ```python
-@rule(StringType, [IntType])
+@rule(str, [int])
 def int_to_str(an_int):
-  return '{}'.format(an_int)
+  return str(an_int)
 ```
 
-The first argument to the `@rule` decorator is the Product (ie, return) type for the `@rule`. The
-second argument is a list of parameter selectors that declare the types of the input parameters for
-the `@rule`. In this case, because the Product type is `StringType` and there is one parameter
-selector (for `IntType`), this `@rule` represents a conversion from `IntType` to `StrType`, with no
-other inputs.
+The first argument to the `@rule` decorator is the `Product` (ie, return) type for the `@rule`. The
+second argument is a list of "parameter selectors" that declare the types of the input parameters for
+the `@rule`. In this case, because the `Product` type is `str` and there is one parameter
+selector (for `int`), this `@rule` represents a conversion from `int` to `str`, with no other inputs.
 
-When the engine statically checks whether it can use this `@rule` to create a string for a
-Subject, it will first see whether there are any ways to get an IntType for that Subject. If
-the subject is already of `type(subject) == IntType`, then the `@rule` will be satisfiable without
-any other dependencies. On the other hand, if the type _doesn't_ match, the engine doesn't give up:
-it will next look for any other registered `@rule`s that can compute an IntType Product for the
-Subject (and so on, recursively).
+When the engine encounters this `@rule` while compiling the rule graph for `str`-producing-`@rules`,
+it will next go hunting for the dependency `@rule` that can produce an `int` using the fewest number
+of `Params`. For example, if there was an `@rule` that could produce an `int` without consuming any
+`Params` at all (ie, a singleton), then that `@rule` would always be chosen first. If all `@rules` to
+produce `int`s required at least one `Param`, then the engine would next see whether the input `Params`
+contained an `int`, or whether there were any `@rules` that required only one `Param`... then two
+`Params`, and etc.
+
+In cases where this search detects any ambiguity (generally because there are two or more `@rules` that
+can provide the same product with the same number of parameters), rule graph compilation will fail with
+a useful error message.
 
 ### Datatypes
 
-In practical use, using basic types like `StringType` or `IntType` does not provide enough
-information to disambiguate between various types of data. So declaring small `datatype`
-definitions to provide a unique and descriptive type is strongly recommended:
+In practical use, builtin types like `str` or `int` do not provide enough information to disambiguate
```

Comment:
Is the issue not providing enough information, or instead you can't easily compose complex data types like a `List[int]` or a `Tuple[MyClass, int]`.

Currently this suggests it's wrong to ever use `str` or `int`, because it suggests they result in ambiguity. Instead, I think what we're after is it would be hard to express complex data types.

Reply:
Hm, well after reading the example maybe both reasons are cause to use a `datatype`. Would be useful to mention both.

Reply:
Yea, it's both. A feature, not a bug =)

Reply:
Perhaps `to create compound types and to disambiguate between various types of data`? I do think the idea of collections is an important one to explain. 

When reading this, I wasn't immediately thinking "huh how would I make a list of ints?" But that's an important thing to know it's possible, so I think it's worth proactively mentioning it.